test/extended/util/workspace.go

```diff
@@ -0,0 +1,30 @@
+package util
+
+import (
+	"fmt"
+
+	o "github.com/onsi/gomega"
+	"k8s.io/apiserver/pkg/storage/names"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+)
+
+// WorkSpace defination
+type WorkSpace struct {
+	Name             string   // WorkSpace Name                                                E.g. e2e-test-kcp-workspace-xxxxx
+	CurrentNameSpace string   // The latest workSpace's namespace created by SetupNameSpace()  E.g. e2e-ns-kcp-workspace-xxxxx
+	Namespaces       []string // WorkSpace's Namespaces created by SetupNameSpace()            E.g. e2e-ns-kcp-workspace-xxxxx
+	ServerURL        string   // WorkSpace ServerURL                                           E.g. https://{{kcp-service-domain}}/clusters/root:orgID:e2e-test-kcp-workspace-xxxxx
+	ParentServerURL  string   // WorkSpace ParentServerURL                                     E.g. https://{{kcp-service-domain}}/clusters/root:orgID
+}
+
+// GenerateNamespace creates a new namespace with
+// name in the format of "e2e-ns-"" + basename + 5Bytes random string
+func (ws *WorkSpace) CreateNamespace(c *CLI) {
```

Comment:
TBH, I still feel that the usage of the method is a bit strange. I think keep the `setupNs` in client.go is better. Let's wait other guys suggestions.

Reply:
Same as previous comment, from my perspective, using value or pointer is technically either fine.
If using pointer, s/GenerateNamespace creates/CreateNamespace creates/.
If using value, s/func ... SetupNamespace()/func ... SetNamespaceInWorkspace()/, otherwise it will be confusing that `SetupNamespace` and `SetNamespace` both exist.

Reply:
Hi @xingxingxia, it's just my opinion, but it seems a bit messy to create a `SetWorkspaceNamespace` function. Also there are benefits to abstracting out the workspace object like readable code and, in the future, we can add more workspace specific helper functions into the `workspace.go` file.

Reply:
OK